pkg/ebpf/telemetry/errors_telemetry.go

```diff
@@ -43,213 +40,346 @@ var helperNames = map[int]string{
+			if sizes[program] > 8 /*bytes*/ {
+				return errors.New("instrumentation block cannot perform any stack reads or writes on more than one stack slot")
+			}
+
+			ins := collectionSpec.Programs[program].Instructions
 
-	// the `probeKeys` get added during instruction patching, so we just try to insert entries for any that don't exist
-	z := new(HelperErrTelemetry)
-	for p, key := range b.probeKeys {
-		err := b.helperErrMap.Update(&key, z, ebpf.UpdateNoExist)
-		if err != nil && !errors.Is(err, ebpf.ErrKeyExist) {
-			return fmt.Errorf("failed to initialize telemetry struct for probe %s", p)
+			iter := ins.Iterate()
+			for iter.Next() {
+				ins := iter.Ins
+				// The final instruction in the instrumentation block is `exit`, which we
+				// do not want.
+				if ins.OpCode.JumpOp() == asm.Exit {
+					break
+				}
+				instrumentationBlockCount += countRawBPFIns(ins)
+
+				// The first instruction has associated func_info btf information. Since
+				// the instrumentation is not a function, the verifier will complain that the number of
+				// `func_info` objects in the BTF do not match the number of loaded programs:
+				// https://elixir.bootlin.com/linux/latest/source/kernel/bpf/verifier.c#L15035
+				// To workaround this we create a new instruction object and give it empty metadata.
+				if iter.Index == 0 {
+					newIns := asm.Instruction{
+						OpCode:   ins.OpCode,
+						Dst:      ins.Dst,
+						Src:      ins.Src,
+						Offset:   ins.Offset,
+						Constant: ins.Constant,
+					}.WithMetadata(asm.Metadata{})
+
+					instrumentationBlock = append(instrumentationBlock, &newIns)
+					continue
+				}
+
+				instrumentationBlock = append(instrumentationBlock, ins)
+			}
+		}
+
+		// absolute jump back to the telemetry patch point to continue normal execution
+		retJumpOffset := trampolinePatchIndex - (insCount + instrumentationBlockCount)
+		newIns := asm.Instruction{
+			OpCode: asm.OpCode(asm.JumpClass).SetJumpOp(asm.Ja),
+			Offset: int16(retJumpOffset),
+		}
+		instrumentationBlock = append(instrumentationBlock, &newIns)
+
+		for _, ins := range instrumentationBlock {
+			p.Instructions = append(p.Instructions, *ins)
 		}
 	}
+
 	return nil
 }
 
 // setupForTelemetry sets up the manager to handle eBPF telemetry.
 // It will patch the instructions of all the manager probes and `undefinedProbes` provided.
 // Constants are replaced for map error and helper error keys with their respective values.
 // This must be called before ebpf-manager.Manager.Init/InitWithOptions
-func setupForTelemetry(m *manager.Manager, options *manager.Options, bpfTelemetry *EBPFTelemetry) error {
-	activateBPFTelemetry, err := ebpfTelemetrySupported()
+func setupForTelemetry(m *manager.Manager, options *manager.Options, bpfTelemetry *EBPFTelemetry, bytecode io.ReaderAt, shouldSkip func(string) bool) error {
+	bpfTelemetry.mtx.Lock()
+	defer bpfTelemetry.mtx.Unlock()
+
+	instrumented, err := ELFBuiltWithInstrumentation(bytecode)
```

Comment:
I think this function will advance the `bytecode` reader

Reply:
Same comment as above. I don't think `io.ReaderAt` advances the cursor.